File: algorithms/graph/connected-component-labeling/swift/ConnectedComponents.swift
import Foundation

class ConnectedComponents {
    static func solve(_ arr: [Int]) -> [Int] {
        if arr.count < 2 { return [] }
        
        let n = arr[0]
        let m = arr[1]
        
        if arr.count < 2 + 2 * m { return [] }
        if n == 0 { return [] }
        
        var adj = [[Int]](repeating: [], count: n)
        for i in 0..<m {
            let u = arr[2 + 2 * i]
            let v = arr[2 + 2 * i + 1]
            if u >= 0 && u < n && v >= 0 && v < n {
                adj[u].append(v)
                adj[v].append(u)
            }
        }
        
        var labels = [Int](repeating: -1, count: n)
        var q = [Int]()
        
        for i in 0..<n {
            if labels[i] == -1 {
                let componentId = i
                labels[i] = componentId
                q.append(i)
                
                var head = 0
                while head < q.count {
                    let u = q[head]
                    head += 1
                    
                    for v in adj[u] {
                        if labels[v] == -1 {
                            labels[v] = componentId
                            q.append(v)
                        }
                    }
                }
                q.removeAll()
            }
        }
        
        return labels
    }
}

File: algorithms/graph/connected-component-labeling/csharp/ConnectedComponents.cs
using System;
using System.Collections.Generic;

namespace Algorithms.Graph.ConnectedComponentLabeling
{
    public class ConnectedComponents
    {
        public static int[] Solve(int[] arr)
        {
            if (arr == null || arr.Length < 2) return new int[0];

            int n = arr[0];
            int m = arr[1];

            if (arr.Length < 2 + 2 * m) return new int[0];
            if (n == 0) return new int[0];

            List<int>[] adj = new List<int>[n];
            for (int i = 0; i < n; i++) adj[i] = new List<int>();

            for (int i = 0; i < m; i++)
            {
                int u = arr[2 + 2 * i];
                int v = arr[2 + 2 * i + 1];
                if (u >= 0 && u < n && v >= 0 && v < n)
                {
                    adj[u].Add(v);
                    adj[v].Add(u);
                }
            }

            int[] labels = new int[n];
            for (int i = 0; i < n; i++) labels[i] = -1;

            Queue<int> q = new Queue<int>();

            for (int i = 0; i < n; i++)
            {
                if (labels[i] == -1)
                {
                    int componentId = i;
                    labels[i] = componentId;
                    q.Enqueue(i);

                    while (q.Count > 0)
                    {
                        int u = q.Dequeue();

                        foreach (int v in adj[u])
                        {
                            if (labels[v] == -1)
                            {
                                labels[v] = componentId;
                                q.Enqueue(v);
                            }
                        }
                    }
                }
            }

            return labels;
        }
    }
}
